File: my-folder/0399-evaluate-division/solution.go
type Point struct {
    Denominator string
    Value float64
}

func calcEquation(equations [][]string, values []float64, queries [][]string) []float64 {
    adj := make(map[string][]Point)

    for i, e := range equations {
        adj[e[0]] = append(adj[e[0]], Point{e[1], values[i]})
        adj[e[1]] = append(adj[e[1]], Point{e[0], 1 / values[i]})
    }

    var bfs func(src, target string) float64
    bfs = func(src, target string) float64 {
        if _, ok := adj[src]; !ok {
            return -1.0
        }
        if _, ok := adj[target]; !ok {
            return -1.0
        }

        q := [][2]interface{}{{src, 1.0}}
        visited := make(map[string]bool)
        visited[src] = true

        for len(q) > 0 {
            node := q[0][0].(string)
            weight := q[0][1].(float64)
            q = q[1:]

            if node == target {
                return weight
            }

            for _, ngh := range adj[node] {
                nghNode := ngh.Denominator
                nghWeight := ngh.Value

                if !visited[nghNode] {
                    visited[nghNode] = true
                    q = append(q, [2]interface{}{nghNode, weight * nghWeight})
                }
            }
        }

        return -1.0
    }

    res := make([]float64, len(queries))
    for i, q := range queries {
        res[i] = bfs(q[0], q[1])
    }

    return res
}

File: my-folder/0399-evaluate-division/solution.rs
use std::collections::{HashMap, HashSet, VecDeque};

struct Edge {
    To: String,
    Weight: f64,
}

impl Solution {
    pub fn calc_equation(equations: Vec<Vec<String>>, values: Vec<f64>, queries: Vec<Vec<String>>) -> Vec<f64> {
        let mut graph: HashMap<String, Vec<Edge>> = HashMap::new();
        for (i, eq) in equations.iter().enumerate() {
            graph.entry(eq[0].clone()).or_insert_with(Vec::new).push(Edge {
                To: eq[1].clone(),
                Weight: values[i]
            });
            graph.entry(eq[1].clone()).or_insert_with(Vec::new).push(Edge {
                To: eq[0].clone(),
                Weight: 1.0 / values[i]
            });
        }
        
        let mut res = vec![-1.0; queries.len()];
        for i in 0..queries.len() {
            res[i] = bfs(&graph, &queries[i][0], &queries[i][1]);
        }

        return res;
    }
}

fn bfs(graph: &HashMap<String, Vec<Edge>>, source: &str, dst: &str) -> f64 {
    if !graph.contains_key(source) || !graph.contains_key(dst) {
        return -1.0;
    }

    let mut queue: VecDeque<(&str, f64)> = VecDeque::new();
    let mut visited: HashSet<&str> = HashSet::new();
    queue.push_back((source, 1.0));
    visited.insert(source);


    while let Some((node, weight)) = queue.pop_front() {
        if node == dst {
            return weight;
        }
        if let Some(edges) = graph.get(node) {
            for edge in edges {
                if !visited.contains(edge.To.as_str()) {
                    visited.insert(&edge.To);
                    queue.push_back((&edge.To, weight*edge.Weight));
                }
            }
        }
    }
    
    -1.0
}
